GVFS/GVFS.Common/Cleanup/GitCleanupStep.cs

```diff
@@ -0,0 +1,198 @@
+﻿using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common.Cleanup
+{
+    public abstract class GitCleanupStep
+    {
+        public const string ObjectCacheLock = "git-cleanup-step.lock";
+        private readonly object gitProcessLock = new object();
+
+        public GitCleanupStep(GVFSContext context, GitObjects gitObjects, bool requireCacheLock = false)
+        {
+            this.Context = context;
+            this.GitObjects = gitObjects;
+            this.RequireCacheLock = requireCacheLock;
+        }
+
+        public abstract string TelemetryKey { get; }
+
+        protected GVFSContext Context { get; }
+        protected GitObjects GitObjects { get; }
+        protected GitProcess GitProcess { get; private set; }
+        protected bool Stopping { get; private set; }
+        protected bool RequireCacheLock { get; }
+
+        public static bool TryStopGitProcess(ITracer tracer, GitProcess process)
+        {
+            if (process == null)
+            {
+                return false;
+            }
+
+            if (process.TryKillRunningProcess())
+            {
+                tracer.RelatedEvent(
+                    EventLevel.Informational,
+                    "Killed background Git process during " + nameof(TryStopGitProcess),
+                    metadata: null);
+                return true;
+            }
+            else
+            {
+                tracer.RelatedEvent(
+                    EventLevel.Informational,
+                    "Failed to kill background Git process during " + nameof(TryStopGitProcess),
+                    metadata: null);
+                return false;
+            }
+        }
+
+        public void Execute()
+        {
+            try
+            {
+                if (this.RequireCacheLock)
+                {
+                    using (FileBasedLock cacheLock = GVFSPlatform.Instance.CreateFileBasedLock(
+                        this.Context.FileSystem,
+                        this.Context.Tracer,
+                        Path.Combine(this.Context.Enlistment.GitObjectsRoot, ObjectCacheLock)))
+                    {
+                        if (!cacheLock.TryAcquireLock())
+                        {
+                            this.Context.Tracer.RelatedInfo(this.TelemetryKey + ": Skipping work since another process holds the lock");

```

Comment:
Yes, a prefetch step could block a post-fetch step. The basic philosophy of this background queue is "best effort" and that these steps never _need_ to run. It's just nice when they do.

If we want to work harder to ensure things run later if the lock is blocking, then we can extend the design later (say, by building a way to reschedule). I do think that is out of scope for this PR as it would be a significant change of behavior.